docs/configuration.rst
Configuration
=============
This library can be configured by environment variables or programmatically
via the :func:`~klempner.config.configure` function.  If the
:func:`~klempner.config.configure` is not called explicitly, then
:func:`~klempner.config.configure_from_environment` will be called implicitly
the first time that :func:`~klempner.url.build_url` is called.

The following grammar snippet for a URL comes from :rfc:`3986`.  The goal of
:func:`klempner.url.build_url` is creating properly formed URLs so it is very
important to understand precisely how the different discovery methods work.
The following sub-sections describe how each discovery method populates the
:token:`authority` portion of the URLs.

.. productionlist::
   URI       : scheme ":" hier-part [ "?" query ] [ "#" fragment ]
   hier-part : "//" authority path-empty
             : path-absolute
             : path-rootless
             : path-empty
   authority : [ userinfo "@" ] host [ ":" port ]

Environment variables
---------------------
The library can be configured based on the environment by calling the
:func:`~klempner.config.configure_from_environment` function explicitly.

.. envvar:: KLEMPNER_DISCOVERY

   Controls the discovery method that the library will used.  The following
   values are understood:

      - :ref:`consul-discovery-method`
      - :ref:`consul-agent-discovery-method`
      - :ref:`environment-discovery-method`
      - :ref:`kubernetes-discovery-method`
      - :ref:`simple-discovery-method`

.. envvar:: CONSUL_AGENT_URL

   Configures the Consul agent URL used by the
   :ref:`consul-agent-discovery-method` method.  Note that the path, query,
   and fragment portions of the URL are ignored.

.. envvar:: CONSUL_DATACENTER

   Configures the datacenter used for Consul-based discovery methods.  This
   variable is required if :envvar:`KLEMPNER_DISCOVERY` is set to
   :ref:`consul-discovery-method`.

.. envvar:: CONSUL_HTTP_TOKEN

   Configures the option authorization token for interacting with the
   Consul HTTP API.  If this environment variable is set, then it is
   sent as a HTTP ``Beaerer`` authorization header.

.. envvar:: KUBERNETES_NAMESPACE

   Configures the name of the Kubernetes namespace used by
   :ref:`kubernetes-discovery-method` to generate URLs.  If this variable is
   not set, the value of ``default`` is used.

URL schemes
-----------
The default scheme for all URLs is ``http``.  If a port number is available
for the configured discovery scheme, then the port number is looked up in
:data:`klempner.config.URL_SCHEME_MAP` and the result is used as the URL
scheme.  The initial content of the mapping contains many of the `IANA
registered schemes`_ as well as a number of other commonly used ones (e.g.,
``postgresql``, ``amqp``).

You can adjust the *port to scheme* mapping to match your needs.  If you
want to disable scheme mapping altogether, simply clear the mapping when
your application initializes:

.. code-block:: python

   klempner.config.URL_SCHEME_MAP.clear()

Use the ``update`` operation if you need to augment the mapping or override
specific entries:

.. code-block:: python

   klempner.config.URL_SCHEME_MAP.update({
      5672: 'rabbitmq',
      15672: 'rabbitmq-admin',
   })

The mapping is a simple :class:`dict` so you can manipulate it using the
standard methods.  It is not cached anywhere in the library implementation
so all modifications are immediately reflected in API calls.

.. _IANA registered schemes: https://www.iana.org/assignments/uri-schemes
   /uri-schemes.xhtml



docs/history.rst
Release history
===============

Next Release
------------
- :compare:`0.0.3...master`
- Add support for using https with consul by replacing ``CONSUL_HTTP_ADDR``
  with :envvar:`CONSUL_AGENT_URL`

0.0.3 (25 May 2019)
-------------------
- :compare:`0.0.2...0.0.3`
- Add support for the :envvar:`CONSUL_HTTP_TOKEN` environment variable.
- Set ``User-Agent`` when sending requests to Consul.
- Rework documentation.

0.0.2 (19 May 2019)
-------------------
- :compare:`0.0.1...0.0.2`
- Add :data:`~klempner.config.URL_SCHEME_MAP` so that URL schemes can be
  easily configured.
- Separate the :ref:`environment-discovery-method` method out from the
  :ref:`simple-discovery-method` method.  Previously the simple method would
  fall back to using environment variables if they were set.

0.0.1 (5 May 2019)
------------------
- :compare:`0.0.0...0.0.1`
- Implement :ref:`simple-discovery-method`, :ref:`consul-discovery-method`,
  :ref:`consul-agent-discovery-method`, :ref:`kubernetes-discovery-method`
  discovery methods.


klempner/config.py
import logging
import os

import requests

from klempner import compat, errors, version

URL_SCHEME_MAP = {
    5672: 'amqp',  # https://www.rabbitmq.com/uri-spec.html
    21: 'ftp',  # https://tools.ietf.org/html/rfc1738
    70: 'gopher',  # https://tools.ietf.org/html/rfc4266
    80: 'http',  # https://tools.ietf.org/html/rfc7230#section-2.7.1
    443: 'https',  # https://tools.ietf.org/html/rfc7230#section-2.7.2
    1344: 'icap',  # https://tools.ietf.org/html/rfc3507#section-4.2
    631: 'ipp',  # https://tools.ietf.org/html/rfc3510
    389: 'ldap',  # https://tools.ietf.org/html/rfc4516
    636: 'ldaps',  # https://tools.ietf.org/html/rfc4516
    # https://docs.mongodb.com/manual/reference/connection-string/
    27017: 'mongodb',
    3306: 'mysql',
    119: 'nntp',  # https://tools.ietf.org/html/rfc5538
    110: 'pop',  # https://tools.ietf.org/html/rfc2384
    # https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
    5432: 'postgresql',
    6379: 'redis',  # https://www.iana.org/assignments/uri-schemes/prov/redis
    873: 'rsync',  # https://tools.ietf.org/html/rfc5781
    554: 'rtsp',  # https://tools.ietf.org/html/rfc7826#section-4.2
    322: 'rtsps',  # https://tools.ietf.org/html/rfc7826#section-4.2
    25: 'smtp',  # https://tools.ietf.org/html/draft-melnikov-smime-msa-to-mda
    161: 'snmp',  # https://tools.ietf.org/html/rfc4088
    22: 'ssh',  # https://tools.ietf.org/html/draft-ietf-secsh-scp-sftp-ssh-uri
    23: 'telnet',  # https://tools.ietf.org/html/rfc4248
    69: 'tftp',  # https://tools.ietf.org/html/rfc3617
    3372: 'tip',  # https://tools.ietf.org/html/rfc2371
    5900: 'vnc',  # https://tools.ietf.org/html/rfc7869
    602: 'xmlrpc.beep',  # https://tools.ietf.org/html/rfc3529#section-5.1
}
"""Mapping of port number to URL scheme.

This dictionary is used to identify the URL scheme based on the port number
when a port number is available.  Users of the library MAY modify the content
of this dictionary **at any time**.

"""


class DiscoveryMethod(object):
    """Available discovery methods."""

    SIMPLE = 'simple'
    """Build URLs using service name as host name."""

    CONSUL = 'consul'
    """Build consul-based service URLs."""

    CONSUL_AGENT = 'consul+agent'
    """Build consul-based service URLs using a consul agent."""

    ENV_VARS = 'environment'
    """Build URLs based on _HOST, _PORT, and _SCHEME environment variables."""

    K8S = 'kubernetes'
    """Build Kubernetes cluster-based service URLs."""

    DEFAULT = SIMPLE

    UNSET = object()
    """The initial library state.

    This can be used to reset the library internals by calling
    :func:`.configure` with `UNSET` as the discovery method.

    """

    AVAILABLE = (CONSUL, CONSUL_AGENT, ENV_VARS, K8S, SIMPLE, UNSET)


_discovery_method = DiscoveryMethod.UNSET
_discovery_parameters = {}


def reset():
    """Reset URL construction parameters."""
    configure(DiscoveryMethod.UNSET)


def ensure_configured():
    """Configure from the environment if currently unconfigured."""
    if _discovery_method is DiscoveryMethod.UNSET:
        configure_from_environment()


def configure(discovery_method, **parameters):
    """Configure the discovery method directly.

    :param discovery_method: method to use
    :param parameters: parameters required for the selected
        method
    :raises: :exc:`klempner.errors.ConfigurationError` if a required
        parameter is not provided

    If `discovery_method` is :attr:`.DiscoveryMethod.UNSET`, then the
    current configuration is cleared and the library is reset to an
    unconfigured state.  This means that the next call to
    :func:`~klempner.url.build_url` will configure the library based
    on the current environment variables.

    """
    from klempner import url  # late import to avoid circular dependency

    logger = logging.getLogger(__package__).getChild('configure')

    def require_parameter(name):
        try:
            return parameters.pop(name)
        except KeyError:
            logger.error('parameter %s is required by discovery method %s',
                         name, discovery_method)
            raise errors.ConfigurationError(name, None)

    logger.debug('configuring for discovery_method %s with parameters=%r',
                 discovery_method, parameters)
    incoming_parameters = {}
    if discovery_method == DiscoveryMethod.CONSUL:
        incoming_parameters['datacenter'] = require_parameter('datacenter')
    elif discovery_method == DiscoveryMethod.CONSUL_AGENT:
        incoming_parameters['datacenter'] = require_parameter('datacenter')
    elif discovery_method == DiscoveryMethod.K8S:
        incoming_parameters['namespace'] = require_parameter('namespace')
    elif discovery_method == DiscoveryMethod.UNSET:
        url._reset_cache()
    elif discovery_method not in DiscoveryMethod.AVAILABLE:
        raise errors.ConfigurationError('discovery_style', discovery_method)

    global _discovery_method, _discovery_parameters
    if discovery_method is DiscoveryMethod.UNSET:
        logger.info('resetting/clearing configuration values')
    elif _discovery_method is DiscoveryMethod.UNSET:
        logger.info('setting discovery method to %r with parameters=%r',
                    discovery_method, incoming_parameters)
    else:
        logger.info(
            'setting discovery method: current_method=%r new_method=%r '
            'parameters=%r', _discovery_method, discovery_method,
            incoming_parameters)
    _discovery_method = discovery_method
    _discovery_parameters.clear()
    _discovery_parameters.update(incoming_parameters)
    if parameters:
        logger.warning(
            'discovery style %s does not accept additional parameters, '
            '%d extra parameters were passed to configure', discovery_method,
            len(parameters))


def configure_from_environment():
    """Set the discovery method from ``$KLEMPNER_DISCOVERY``.

    This method configures the library based on the
    :envvar:`KLEMPNER_DISCOVERY` environment variable.  If the
    environment variable is not set, then the "simple" configuration
    is used.

    """
    logger = logging.getLogger(__package__).getChild(
        'configure_from_environment')
    new_method = os.environ.get('KLEMPNER_DISCOVERY', DiscoveryMethod.SIMPLE)
    logger.debug('configuring from environment: discovery_method=%s',
                 new_method)

    def require_envvar(name):
        try:
            return os.environ[name]
        except KeyError:
            logger.error(
                'discovery method %s requires the %s environment variable',
                new_method, name)
            raise errors.ConfigurationError(name, None)

    parameters = {}
    if new_method == DiscoveryMethod.CONSUL:
        parameters['datacenter'] = require_envvar('CONSUL_DATACENTER')
    elif new_method == DiscoveryMethod.CONSUL_AGENT:
        headers = {'User-Agent': '/'.join([__package__, version])}
        try:
            headers['Authorization'] = 'Bearer {}'.format(
                os.environ['CONSUL_HTTP_TOKEN'])
        except KeyError:
            pass
        parsed = compat.urlparse(require_envvar('CONSUL_AGENT_URL'))
        url = compat.urlunparse((parsed[0], parsed[1], '/v1/agent/self', None,
                                 None, None))
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        body = response.json()
        parameters['datacenter'] = body['Config']['Datacenter']
    elif new_method == DiscoveryMethod.K8S:
        parameters['namespace'] = os.environ.get('KUBERNETES_NAMESPACE',
                                                 'default')
    elif new_method not in DiscoveryMethod.AVAILABLE:
        raise errors.ConfigurationError('discovery_style', new_method)

    configure(new_method, **parameters)


def get_discovery_details():
    """Retrieve the configured method and parameters.

    :rtype: tuple(str, dict)

    """
    return _discovery_method, _discovery_parameters.copy()
